lib/src/Pos/Util/Servant.hs

```diff
@@ -229,16 +247,39 @@ instance (ApiHasArgClass apiType a, ApiArg apiType a ~ Maybe b) =>
     apiArgName _ = apiArgName (Proxy @(apiType a))
 
 instance ( HasServer (apiType a :> res) ctx
+         , HasServer res ctx
+         , ApiHasArg (WithDefaultApiArg apiType) a res
          , Server (apiType a :> res) ~ (Maybe c -> d)
          , Default c
          ) =>
          HasServer (WithDefaultApiArg apiType a :> res) ctx where
     type ServerT (WithDefaultApiArg apiType a :> res) m =
          UnmaybeArg (ServerT (apiType a :> res) m)
+
     route =
         inRouteServer @(apiType a :> res) route $
         \f a -> f $ fromMaybe def a
 
+    -- The simpliest what we can do here is to delegate to
+    -- @hoistServer (Proxy @res)@.
+    -- However to perform that we need a knowledge of that
+    -- @ServerT (withDefaultApiArg apiType a :> res) m@ is a /function/
+    -- regardless of @m@.
+    -- I.e. we have to set a constraint
+    -- @forall m. ServerT (WithDefaultApiArg apiType :> res) m ~ (e -> d)@
+    -- (for some @e@ and @d@).
+    -- However GHC doesn't allow @forall@ in constraints.
+    --
+    -- But we can use 'Forall' thing instead.
+    -- @Forall (ApiHasArgInvariant (WithDefaultApiArg apiType) a res)@
+    -- constraint is already provided here (see 'ApiHasArg' definition),
+    -- so we can instantiate it's quantification parameter @m@ to concrete
+    -- @m1@ and @m2@ required by this function, getting desired constraints.
+    hoistServerWithContext _ pc (nt :: forall x. m1 x -> m2 x) s =
```

Comment:
Excellent explanation, although I have to admit it's quite mind-blowing and it makes my head hurt 😉